wallet-new/src/Cardano/Wallet/API/Request/Sort.hs

```diff
@@ -0,0 +1,146 @@
+data SortBy (sym :: [Symbol]) (r :: *) deriving Typeable
+
+-- | The direction for the sort operation.
+data SortDirection =
+      SortAscending
+    | SortDescending
+    deriving Show
+
+-- | A sort operation on an index @ix@ for a resource 'a'.
+data SortOperation ix a =
+      SortByIndex SortDirection (Proxy ix)
+    -- ^ Standard sort by index (e.g. sort_on=balance).
+    | SortIdentity
+    -- ^ Do not perform sorting on this resource.
+
+instance Show (SortOperations a) where
+    show = show . flattenSortOperations
+
+-- | A "bag" of sort operations, where the index constraint are captured in
+-- the inner closure of 'SortOp'.
+data SortOperations a where
+    NoSorts  :: SortOperations a
+    SortOp   :: (Indexable' a, IsIndexOf' a ix, ToIndex a ix)
+             => SortOperation ix a
+             -> SortOperations a
+             -> SortOperations a
+
+instance Show (SortOperation ix a) where
+    show (SortByIndex dir _) = "SortByIndex[" <> show dir <> "]"
+    show SortIdentity        = "SortIdentity"
+
+-- | Handy helper function to show opaque 'FilterOperation'(s), mostly for
+-- debug purposes.
+flattenSortOperations :: SortOperations a -> [String]
+flattenSortOperations NoSorts       = mempty
+flattenSortOperations (SortOp f fs) = show f : flattenSortOperations fs
+
+-- | This is a slighly boilerplat-y type family which maps symbols to
+-- indices, so that we can later on reify them into a list of valid indices.
+-- In case we want to sort on _all_ the indices, it might make sense having an
+-- entry like the following:
+--
+--    SortParams '["wallet_id", "balance"] Wallet = IndicesOf Wallet
+--
+-- In the case of a 'Wallet', for example, sorting by @wallet_id@ doesn't have
+-- much sense, so we restrict ourselves.
+type family SortParams (syms :: [Symbol]) (r :: *) :: [*] where
+    SortParams '["balance"] Wallet = '[Core.Coin]
+
+-- | Handy typeclass to reconcile type and value levels by building a list of 'SortOperation' out of
+-- a type level list.
+class ToSortOperations (ixs :: [*]) a where
+  toSortOperations :: Request -> [Text] -> proxy ixs -> SortOperations a
+
+instance Indexable' a => ToSortOperations ('[]) a where
+  toSortOperations _ _ _ = NoSorts
+
+instance ( Indexable' a
+         , IsIndexOf' a ix
+         , ToIndex a ix
+         , ToSortOperations ixs a
+         )
+         => ToSortOperations (ix ': ixs) a where
+  toSortOperations req [] _     =
+      let newOp = SortIdentity
+      in SortOp (newOp :: SortOperation ix a) (toSortOperations req [] (Proxy :: Proxy ixs))
+  toSortOperations req (key:xs) _ =
+      case List.lookup "sort_by" (parseQueryText $ rawQueryString req) of
+          Nothing       ->
+              let newOp = SortIdentity
+              in SortOp (newOp :: SortOperation ix a) (toSortOperations req xs (Proxy @ ixs))
+          Just Nothing  ->
+              let newOp = SortIdentity
+              in SortOp (newOp :: SortOperation ix a) (toSortOperations req xs (Proxy @ ixs))
```

Comment:
there is some redundacy here with L99

Reply:
This part was copied and adapted from the `servant` codebase -- open a PR upstream! 😁 :trollface: